codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -317,85 +315,98 @@ private Http2Stream requireStream(int streamId) {
+            while (!bufsAndPromises.isEmpty()) {
+                Object bufOrPromise = bufsAndPromises.remove();
+                if (bufOrPromise instanceof ByteBuf) {
+                    ReferenceCountUtil.safeRelease(bufOrPromise);
+                } else {
+                    ((ChannelPromise) bufOrPromise).tryFailure(cause);
+                }
+            }
             size = 0;
-            promise.tryFailure(cause);
+            padding = 0;
+            lifecycleManager.onException(ctx, cause);
         }
 
         @Override
         public void write(int allowedBytes) {
             int bytesWritten = 0;
-            if (data == null || (allowedBytes == 0 && size != 0)) {
+            if (bufsAndPromises.isEmpty() || (allowedBytes == 0 && size != 0)) {
                 // No point writing an empty DATA frame, wait for a bigger allowance.
                 return;
             }
-            try {
-                int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
-                do {
-                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
-                    ByteBuf toWrite;
-                    // Let data consume the frame before padding.
-                    int writeableData = data.readableBytes();
-                    if (writeableData > allowedFrameSize) {
-                        writeableData = allowedFrameSize;
-                        toWrite = data.readSlice(writeableData).retain();
-                    } else {
-                        // We're going to write the full buffer which will cause it to be released, for subsequent
-                        // writes just use empty buffer to avoid over-releasing. Have to use an empty buffer
-                        // as we may continue to write padding in subsequent frames.
-                        toWrite = data;
-                        data = Unpooled.EMPTY_BUFFER;
-                    }
-                    int writeablePadding = Math.min(allowedFrameSize - writeableData, padding);
-                    padding -= writeablePadding;
-                    bytesWritten += writeableData + writeablePadding;
-                    ChannelPromise writePromise;
-                    if (size == bytesWritten && !promise.isVoid()) {
-                        // Can use the original promise if it's the last write
-                        writePromise = promise;
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            do {
+                int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+
+                // Always need a real promise for write as we have to listen to it for failure.
+                ChannelPromise promise = ctx.newPromise();
+                promise.addListener(this);
+                CompositeByteBuf toWrite = ctx.alloc().compositeBuffer(bufsAndPromises.size());
+                while (!bufsAndPromises.isEmpty()) {
```

Comment:
Seems like you are unconditionally removing the first entry.  What about restructuring the loop a bit:

```java
for(;;) {
  final Object entry = bufsAndPromises.pop();
  if (entry == null) {
    break;
  }

  // use entry.
}